engine/test/e2e/e2e_dm_test.go

```diff
@@ -177,154 +185,269 @@ func testSimpleAllModeTask(
+	binlogResp, err = getBinlogOperator(ctx, httpClient, resp.JobId, source1, "", t)
+	require.NoError(t, err)
+	require.Equal(t, "", binlogResp.ErrorMsg)
+	require.Len(t, binlogResp.Results, 1)
+	require.Equal(t, "", binlogResp.Results[source1].Msg)
+	require.Equal(t, fmt.Sprintf("source '%s' has no error", source1), binlogResp.Results[source1].ErrorMsg)
+
+	// binlog delete
+	binlogResp, err = deleteBinlogOperator(ctx, httpClient, resp.JobId, source1, t)
 	require.NoError(t, err)
-	require.Nil(t, resp2.Err)
-	require.NoError(t, json.Unmarshal([]byte(resp2.JsonRet), &binlogResp))
 	require.Equal(t, "", binlogResp.ErrorMsg)
 	require.Len(t, binlogResp.Results, 1)
 	require.Equal(t, "", binlogResp.Results[source1].Msg)
 	require.Equal(t, fmt.Sprintf("source '%s' has no error", source1), binlogResp.Results[source1].ErrorMsg)
-}
 
-func queryStatus(ctx context.Context, client client.MasterClient, jobID string, tasks []string, t *testing.T) (*enginepb.DebugJobResponse, error) {
-	var args struct {
-		Tasks []string
+	// pause task
+	err = operateJob(ctx, httpClient, resp.JobId, []string{source1}, dmpkg.Pause, t)
+	require.NoError(t, err)
+	// eventually paused
+	require.Eventually(t, func() bool {
+		jobStatus, err = queryStatus(ctx, httpClient, resp.JobId, nil, t)
+		require.NoError(t, err)
+		return jobStatus.TaskStatus[source1].Status.Stage == metadata.StagePaused
+	}, time.Second*10, time.Second)
+	// set binlog schema
+	fromSource := true
+	binlogSchemaReq := &openapi.SetBinlogSchemaRequest{
+		Database:   db,
+		Table:      "t1",
+		FromSource: &fromSource,
 	}
-	args.Tasks = tasks
-	jsonArg, err := json.Marshal(args)
+	binlogSchemaResp, err = setBinlogSchema(ctx, httpClient, resp.JobId, source1, binlogSchemaReq, t)
 	require.NoError(t, err)
-	ctx2, cancel := context.WithTimeout(ctx, 30*time.Second)
-	defer cancel()
-	return client.DebugJob(ctx2, &enginepb.DebugJobRequest{JobId: jobID, Command: dmpkg.QueryStatus, JsonArg: string(jsonArg)})
+	require.Equal(t, "", binlogSchemaResp.ErrorMsg)
+	require.Equal(t, "", binlogSchemaResp.Results[source1].ErrorMsg)
+	require.Equal(t, "", binlogSchemaResp.Results[source1].Msg)
+	// get new binlog schema
+	binlogSchemaResp, err = getBinlogSchema(ctx, httpClient, resp.JobId, source1, db, "t1", t)
+	require.NoError(t, err)
+	require.Len(t, binlogSchemaResp.Results, 1)
+	require.Equal(t, "CREATE TABLE `t1` ( `c` int(11) NOT NULL, `new_col` int(11) DEFAULT NULL, PRIMARY KEY (`c`) /*T![clustered_index] CLUSTERED */, UNIQUE KEY `new_col` (`new_col`)) ENGINE=InnoDB DEFAULT CHARSET=latin1 COLLATE=latin1_bin", binlogSchemaResp.Results[source1].Msg)
+	require.Equal(t, "", binlogSchemaResp.Results[source1].ErrorMsg)
+}
+
+func queryStatus(ctx context.Context, client *http.Client, jobID string, tasks []string, t *testing.T) (*dm.JobStatus, error) {
+	url := fmt.Sprintf(baseURL+"/status", jobID)
+	for i, task := range tasks {
+		if i == 0 {
+			url += "?"
+		} else {
+			url += "&"
+		}
+		url += "tasks=" + task
```

Comment:
maybe https://pkg.go.dev/net/url#Values.Encode because it also handle special characters encoding